packages/core/src/directive/directive.ts

```diff
@@ -20,13 +22,37 @@ export class Directive<T extends keyof DirectiveArgs = any> {
   params(): Param[] {
     if (!this.hasArgs()) return [];
 
+    if (isParamMap(this.args)) {
+      return Object.values(this.args)
+    }
+
     return Array.isArray(this.args)
+        // Assumes args is a array of objects which have a "params" method which returns Param[]
       ? this.args.reduce((r, x) => [...r, ...x.params()], [])
+       // Assumes args has a "params" method which returns Param[]
+       // @ts-expect-error
       : this.args.params();
+
+    // TODO: The assumptions above are not type safe. We need an interface that defines a params() method which returns Param[].
+    // this.args should then be contrained by a type which unites the above and ParamMap.
   }
 
   toString() {
-    return '@' + this.name
-      + (this.hasArgs() ? `(${this.args})` : '');
+    const out = '@' + this.name
+
+    if (!this.hasArgs()) {
+      return out
+    }
+
+    if (isParamMap(this.args)) {
```

Comment:
please note that `this.args` might contain partial params. Or lets say user calls: `query().recurse({ loop: true, depth: params.int(5) })`, then depth param will be ignored since `isParamMap` checks if all the args is paramMap.

Haven't tested the following but should work maybe with slight modification:
```typescript
const argsList = !this.hasArgs() ? []
  : Array.isArray(this.args) ? this.args
  : Object.entries(this.args).map(([key, value]) => `${key}: ${value}`);
return `@${this.name}${
  !argsList.length ? '' : `(${argsList.join(', ')})`
}`;
```
Probably we should change:
https://github.com/binier/dgraphium/blob/641005380c90fa3537223498e4e7eac7d150b45b/packages/core/src/directive/directive-builder.ts#L7
to:
`filter: [LogicalOperatorBuilder | OperatorBuilder]` (and to other places for **filter** directive)
to have the clear distinction between positional and named arguments. Note that without this change code that I wrote above won't work.

Reply:
Since this supports existing feature which is it's not entirely clear to me, can you do this?